### ISSUE TITLE
[Legacy] Port NaN-safe observation operator from mfourdvar

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/fourdvarjax#16]
## Context

`mfourdvar`'s `ObsOperator` ([`mfourdvar/_src/operators/base.py`](https://github.com/jejjohnson/mfourdvar/blob/main/mfourdvar/_src/operators/base.py)) applies `jnp.nan_to_num` before computing the observation loss. This is critical for real-world geophysical data (e.g., satellite altimetry) where observations contain NaN-filled gaps rather than explicit binary masks.

### Legacy source
```python
# mfourdvar/_src/operators/base.py
def loss(self, x, y, mask=None, params=None):
    y_pred = self(x, params=params)
    y = jnp.nan_to_num(y)   # ← NaN-safe
    if mask is not None:
        loss = jnp.sum(mask * (y_pred - y) ** 2)
    else:
        loss = jnp.sum((y_pred - y) ** 2)
    return loss
```

### Current `fourdvarjax` gap
`obs_cost_1d` and `obs_cost_2d` in `fourdvarjax/_src/costs.py` assume a clean binary mask and do not handle NaN inputs.

## Acceptance Criteria
- [ ] `obs_cost_1d_nan` and `obs_cost_2d_nan` variants added to `fourdvarjax/_src/costs.py`
- [ ] Alternatively, `obs_cost_1d` accepts an optional `nan_to_num: bool = False` kwarg
- [ ] Docstrings explain when to use NaN-safe vs. mask-only variants
- [ ] Unit tests with NaN-containing observation arrays
- [ ] Exported from `fourdvarjax/__init__.py`

## Depends on
- None (standalone addition)
